# Supply Chain Analytics
## Preprocsessing

In [1]:
import pandas as pd

### Load and Read Dataset

In [2]:
supply_data = pd.read_csv("cosmetics_supply_chain_data.csv")
supply_data.head()

Product type   SKU      Price  Availability  Number of products sold  \
0     haircare  SKU0  69.808006            55                      802   
1     skincare  SKU1  14.843523            95                      736   
2     haircare  SKU2  11.319683            34                        8   
3     skincare  SKU3  61.163343            68                       83   
4     skincare  SKU4   4.805496            26                      871   

   Revenue generated Customer demographics  Stock levels  Lead times  \
0        8661.996792            Non-binary            58           7   
1        7460.900065                Female            53          30   
2        9577.749626               Unknown             1          10   
3        7766.836426            Non-binary            23          13   
4        2686.505152            Non-binary             5           3   

   Order quantities  ...  Location Lead time  Production volumes  \
0                96  ...    Mumbai        29                 215   
1                37  ...    Mumbai        23                 517   
2                88  ...    Mumbai        12                 971   
3                59  ...   Kolkata        24                 937   
4                56  ...     Delhi         5                 414   

  Manufacturing lead time Manufacturing costs  Inspection results  \
0                      29           46.279879             Pending   
1                      30           33.616769             Pending   
2                      27           30.688019             Pending   
3                      18           35.624741                Fail   
4                       3           92.065161                Fail   

   Defect rates  Transportation modes   Routes       Costs  
0      0.226410                  Road  Route B  187.752075  
1      4.854068                  Road  Route B  503.065579  
2      4.580593                   Air  Route C  141.920282  
3      4.746649                  Rail  Route A  254.776159  
4      3.145580                   Air  Route A  923.440632  

[5 rows x 24 columns]

**Problem Statement:** Optimize the Makeup Product Supply Chain to Maximize Efficiency and Customer Satisfaction

**We can choose any of these objectives:**

*Inventory Optimization*: Minimize stockouts and overstocking of Makeup products to ensure optimal inventory levels. This will help reduce carrying costs and increase revenue.

*Supply Chain Efficiency*: Optimize lead times, production volumes, and transportation routes to streamline the supply chain process. This should lead to faster order fulfillment and reduced shipping costs.

*Customer Satisfaction*: Analyze customer demographics and feedback to understand customer preferences and expectations. Use this information to tailor the supply chain process to meet customer demands effectively.

*Cost Reduction*: Identify areas of cost inefficiencies in the supply chain, such as high manufacturing costs, shipping expenses, or defect rates. Implement measures to reduce these costs while maintaining product quality.

*Supplier Relationship Management*: Evaluate supplier performance and lead times to ensure smooth collaboration with reliable suppliers and minimize delays.

**Demand Forecasting:** Use historical sales data and customer trends to forecast future demand for Cosmetic products accurately. This will help in proactive planning and inventory management.

*Target Variable*: The target variable for demand forecasting is 'Number of products sold'. This is the variable we want to predict based on historical data and other relevant features.

*Relevant Features*: To predict the number of products sold, the following features could be relevant:

'Price': The price of the product can influence demand. Higher prices might result in lower demand, and vice versa.

'Availability': The availability of the product in stock can affect sales. Higher availability may lead to higher sales.

'Customer demographics': Customer characteristics such as gender, age, or location might impact product demand.

'Stock levels': The current stock levels of the product can influence demand. Lower stock levels might lead to increased sales due to scarcity.

'Lead times': The time it takes to fulfill an order can impact demand. Longer lead times might result in lower customer satisfaction and reduced sales.

*Other relevant features*: Depending on the domain and specific context, other features like marketing campaigns, seasonal factors, or competitor data might also be relevant.

In [3]:
# Data shape
supply_data.shape

(100, 24)

In [4]:
# Data Information
supply_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Product type             100 non-null    object 
 1   SKU                      100 non-null    object 
 2   Price                    100 non-null    float64
 3   Availability             100 non-null    int64  
 4   Number of products sold  100 non-null    int64  
 5   Revenue generated        100 non-null    float64
 6   Customer demographics    100 non-null    object 
 7   Stock levels             100 non-null    int64  
 8   Lead times               100 non-null    int64  
 9   Order quantities         100 non-null    int64  
 10  Shipping times           100 non-null    int64  
 11  Shipping carriers        100 non-null    object 
 12  Shipping costs           100 non-null    float64
 13  Supplier name            100 non-null    object 
 14  Location                 10

In [5]:
# Numeric Data Description
supply_data.describe()

Price  Availability  Number of products sold  Revenue generated  \
count  100.000000    100.000000               100.000000         100.000000   
mean    49.462461     48.400000               460.990000        5776.048187   
std     31.168193     30.743317               303.780074        2732.841744   
min      1.699976      1.000000                 8.000000        1061.618523   
25%     19.597823     22.750000               184.250000        2812.847151   
50%     51.239831     43.500000               392.500000        6006.352023   
75%     77.198228     75.000000               704.250000        8253.976921   
max     99.171329    100.000000               996.000000        9866.465458   

       Stock levels  Lead times  Order quantities  Shipping times  \
count    100.000000  100.000000        100.000000      100.000000   
mean      47.770000   15.960000         49.220000        5.750000   
std       31.369372    8.785801         26.784429        2.724283   
min        0.000000    1.000000          1.000000        1.000000   
25%       16.750000    8.000000         26.000000        3.750000   
50%       47.500000   17.000000         52.000000        6.000000   
75%       73.000000   24.000000         71.250000        8.000000   
max      100.000000   30.000000         96.000000       10.000000   

       Shipping costs   Lead time  Production volumes  \
count      100.000000  100.000000          100.000000   
mean         5.548149   17.080000          567.840000   
std          2.651376    8.846251          263.046861   
min          1.013487    1.000000          104.000000   
25%          3.540248   10.000000          352.000000   
50%          5.320534   18.000000          568.500000   
75%          7.601695   25.000000          797.000000   
max          9.929816   30.000000          985.000000   

       Manufacturing lead time  Manufacturing costs  Defect rates       Costs  
count                100.00000           100.000000    100.000000  100.000000  
mean                  14.77000            47.266693      2.277158  529.245782  
std                    8.91243            28.982841      1.461366  258.301696  
min                    1.00000             1.085069      0.018608  103.916248  
25%                    7.00000            22.983299      1.009650  318.778455  
50%                   14.00000            45.905622      2.141863  520.430444  
75%                   23.00000            68.621026      3.563995  763.078231  
max                   30.00000            99.466109      4.939255  997.413450

Key Points from the Table:
1. Stock Levels
* Target Variable: Stock levels.
* Average (Mean): 47.77.
* Minimum: 0 (indicates potential inventory shortage).
* Maximum: 100.0.
* Range: Very wide, indicating high dispersion.
* Median (Q2): 47.50, nearly equal to the mean, suggesting normal (symmetric) distribution.
2. Route Costs
* Target Variable: Route costs.
* Average (Mean): 529.25.
* Minimum: 103.92.
* Maximum: 997.41.
* Range: Very wide.
* Median (Q2): 520.43, close to the mean, indicating normal (symmetric) distribution.

3. We'll check other targets later

In [6]:
# Columns names
supply_data.columns

Index(['Product type', 'SKU', 'Price', 'Availability',
       'Number of products sold', 'Revenue generated', 'Customer demographics',
       'Stock levels', 'Lead times', 'Order quantities', 'Shipping times',
       'Shipping carriers', 'Shipping costs', 'Supplier name', 'Location',
       'Lead time', 'Production volumes', 'Manufacturing lead time',
       'Manufacturing costs', 'Inspection results', 'Defect rates',
       'Transportation modes', 'Routes', 'Costs'],
      dtype='object')

Refer to: [Feature Explanation](cosmetics_supply_chain_data_card.txt)

### Data pre-processing and cleaning

In [7]:
# Check for missing values
missing_values = supply_data.isnull().sum()
missing_values

Product type               0
SKU                        0
Price                      0
Availability               0
Number of products sold    0
Revenue generated          0
Customer demographics      0
Stock levels               0
Lead times                 0
Order quantities           0
Shipping times             0
Shipping carriers          0
Shipping costs             0
Supplier name              0
Location                   0
Lead time                  0
Production volumes         0
Manufacturing lead time    0
Manufacturing costs        0
Inspection results         0
Defect rates               0
Transportation modes       0
Routes                     0
Costs                      0
dtype: int64

**No missing values in the data**

In [ ]:
# Check for duplicates
supply_data.duplicated().any()

False

**No duplicates in the data**

#### Change columns names

In [3]:
supply_data.columns = [col.lower().replace(' ', '_') for col in supply_data.columns]
supply_data.rename(columns=lambda x: x.replace("(", "").replace(")", ""), inplace=True)
supply_data.columns

Index(['product_type', 'sku', 'price', 'availability',
       'number_of_products_sold', 'revenue_generated', 'customer_demographics',
       'stock_levels', 'lead_times', 'order_quantities', 'shipping_times',
       'shipping_carriers', 'shipping_costs', 'supplier_name', 'location',
       'lead_time', 'production_volumes', 'manufacturing_lead_time',
       'manufacturing_costs', 'inspection_results', 'defect_rates',
       'transportation_modes', 'routes', 'costs'],
      dtype='object')

#### Save Cleaned Data

In [5]:
# Unhash when needed
#supply_data.to_csv('cosmetics_supply_data_cleaned.csv', index=False)

#### Check outliers in numerical variables

In [11]:
df_numeric = supply_data[["price", "availability", "number_of_products_sold", "revenue_generated", "stock_levels", "lead_times", "order_quantities", "shipping_times", "shipping_costs", "lead_time", "production_volumes", "manufacturing_lead_time", "manufacturing_costs", "defect_rates", "costs"]]
df_numeric

price  availability  number_of_products_sold  revenue_generated  \
0   69.808006            55                      802        8661.996792   
1   14.843523            95                      736        7460.900065   
2   11.319683            34                        8        9577.749626   
3   61.163343            68                       83        7766.836426   
4    4.805496            26                      871        2686.505152   
..        ...           ...                      ...                ...   
95  77.903927            65                      672        7386.363944   
96  24.423131            29                      324        7698.424766   
97   3.526111            56                       62        4370.916580   
98  19.754605            43                      913        8525.952560   
99  68.517833            17                      627        9185.185829   

    stock_levels  lead_times  order_quantities  shipping_times  \
0             58           7                96               4   
1             53          30                37               2   
2              1          10                88               2   
3             23          13                59               6   
4              5           3                56               8   
..           ...         ...               ...             ...   
95            15          14                26               9   
96            67           2                32               3   
97            46          19                 4               9   
98            53           1                27               7   
99            55           8                59               6   

    shipping_costs  lead_time  production_volumes  manufacturing_lead_time  \
0         2.956572         29                 215                       29   
1         9.716575         23                 517                       30   
2         8.054479         12                 971                       27   
3         1.729569         24                 937                       18   
4         3.890548          5                 414                        3   
..             ...        ...                 ...                      ...   
95        8.630339         18                 450                       26   
96        5.352878         28                 648                       28   
97        7.904846         10                 535                       13   
98        1.409801         28                 581                        9   
99        1.311024         29                 921                        2   

    manufacturing_costs  defect_rates       costs  
0             46.279879      0.226410  187.752075  
1             33.616769      4.854068  503.065579  
2             30.688019      4.580593  141.920282  
3             35.624741      4.746649  254.776159  
4             92.065161      3.145580  923.440632  
..                  ...           ...         ...  
95            58.890686      1.210882  778.864241  
96            17.803756      3.872048  188.742141  
97            65.765156      3.376238  540.132423  
98             5.604691      2.908122  882.198864  
99            38.072899      0.346027  210.743009  

[100 rows x 15 columns]

In [12]:
df_zscores = (df_numeric - df_numeric.mean())/df_numeric.std()
df_zscores

price  availability  number_of_products_sold  revenue_generated  \
0   0.652766      0.214681                 1.122556           1.056025   
1  -1.110714      1.515777                 0.905293           0.616520   
2  -1.223773     -0.468394                -1.491177           1.391117   
3   0.375411      0.637537                -1.244288           0.728468   
4  -1.432774     -0.728614                 1.349694          -1.130524   
..       ...           ...                      ...                ...   
95  0.912516      0.539955                 0.694614           0.589246   
96 -0.803362     -0.631031                -0.450951           0.703435   
97 -1.473821      0.247208                -1.313417          -0.514165   
98 -0.953147     -0.175648                 1.487951           1.006244   
99  0.611372     -1.021360                 0.546481           1.247470   

    stock_levels  lead_times  order_quantities  shipping_times  \
0       0.326114   -1.019827          1.746537       -0.642371   
1       0.166723    1.598033         -0.456235       -1.376509   
2      -1.490945   -0.678367          1.447856       -1.376509   
3      -0.789624   -0.336907          0.365138        0.091767   
4      -1.363432   -1.475107          0.253132        0.825905   
..           ...         ...               ...             ...   
95     -1.044650   -0.223087         -0.866922        1.192975   
96      0.613018   -1.588927         -0.642911       -1.009440   
97     -0.056424    0.346013         -1.688294        1.192975   
98      0.166723   -1.702747         -0.829586        0.458836   
99      0.230480   -0.906007          0.365138        0.091767   

    shipping_costs  lead_time  production_volumes  manufacturing_lead_time  \
0        -0.977446   1.347463           -1.341358                 1.596646   
1         1.572175   0.669210           -0.193274                 1.708849   
2         0.945294  -0.574255            1.532655                 1.372241   
3        -1.440226   0.782252            1.403400                 0.362415   
4        -0.625185  -1.365550           -0.584839                -1.320627   
..             ...        ...                 ...                      ...   
95        1.162487   0.103999           -0.447981                 1.260038   
96       -0.073649   1.234421            0.304737                 1.484444   
97        0.888858  -0.800339           -0.124845                -0.198599   
98       -1.560831   1.234421            0.050029                -0.647410   
99       -1.598086   1.347463            1.342574                -1.432830   

    manufacturing_costs  defect_rates     costs  
0             -0.034048     -1.403309 -1.322073  
1             -0.470966      1.763358 -0.101355  
2             -0.572017      1.576221 -1.499508  
3             -0.401684      1.689851 -1.062593  
4              1.545689      0.594253  1.526102  
..                  ...           ...       ...  
95             0.401065     -0.729643  0.966383  
96            -1.016565      1.091369 -1.318240  
97             0.638256      0.752091  0.042147  
98            -1.437471      0.431763  1.366437  
99            -0.317215     -1.321456 -1.233065  

[100 rows x 15 columns]

#### Z-score

In [13]:
def filter_column_by_threshold(data: pd.DataFrame, column: str, threshold: float) -> pd.DataFrame:
    """
    Filters rows in a DataFrame based on a threshold applied to a specific column.

    Parameters:
    - data (pd.DataFrame): The input DataFrame containing the data.
    - column (str): The name of the column to apply the threshold filter on.
    - threshold (float): The threshold value; rows where the column's value exceeds
                         this threshold will be kept.

    Returns:
    - pd.DataFrame: A DataFrame containing only the rows where the column's value
                    is greater than the specified threshold.
    """
    # Filter the DataFrame to keep only rows where the column's value is above the threshold
    filtered_data = data[data[column] > threshold]
    
    return filtered_data

In [15]:
filtered_data = filter_column_by_threshold(df_zscores, "price", 3)
print(filtered_data)

Empty DataFrame
Columns: [price, availability, number_of_products_sold, revenue_generated, stock_levels, lead_times, order_quantities, shipping_times, shipping_costs, lead_time, production_volumes, manufacturing_lead_time, manufacturing_costs, defect_rates, costs]
Index: []


**No Outliers**